python/lsst/daf/butler/registry/_registry.py

```diff
@@ -1032,6 +1033,72 @@ def insertDatasetLocations(self, datastoreName: str, refs: Iterable[DatasetRef])
             *[{"datastore_name": datastoreName, "dataset_id": _checkAndGetId(ref)} for ref in refs]
         )
 
+    @transactional
+    def moveDatasetLocationToTrash(self, datastoreName: str, refs: Iterable[DatasetRef]):
+        """Move the dataset location information to trash.
+
+        Parameters
+        ----------
+        datastoreName : `str`
+            Name of the datastore holding these datasets.
+        refs : `~collections.abc.Iterable` of `DatasetRef`
+            References to the datasets.
+        """
+        # We only want to move rows that already exist in the main table
+        filtered = self.checkDatasetLocations(datastoreName, refs)
+        self.canDeleteDatasetLocations(datastoreName, filtered)
+        self.removeDatasetLocation(datastoreName, filtered)
+
+    @transactional
+    def canDeleteDatasetLocations(self, datastoreName: str, refs: Iterable[DatasetRef]):
+        """Record that a datastore can delete this dataset
+
+        Parameters
+        ----------
+        datastoreName : `str`
+            Name of the datastore holding these datasets.
+        refs : `~collections.abc.Iterable` of `DatasetRef`
+            References to the datasets.
+
+        Raises
+        ------
+        AmbiguousDatasetError
+            Raised if ``any(ref.id is None for ref in refs)``.
+        """
+        self._db.insert(
+            self._tables.dataset_location_trash,
+            *[{"datastore_name": datastoreName, "dataset_id": _checkAndGetId(ref)} for ref in refs]
+        )
+
+    def checkDatasetLocations(self, datastoreName: str, refs: Iterable[DatasetRef]) -> List[DatasetRef]:
+        """Check which refs are listed for this datastore.
+
+        Parameters
+        ----------
+        datastoreName : `str`
+            Name of the datastore holding these datasets.
+        refs : `~collections.abc.Iterable` of `DatasetRef`
+            References to the datasets.
+
+        Returns
+        -------
+        present : `list` of `DatasetRef`
+            All the `DatasetRef` that are listed.
+        """
+
+        table = self._tables.dataset_location
+        result = self._db.query(
+            sqlalchemy.sql.select(
+                [table.columns.datastore_name, table.columns.dataset_id]
+            ).where(
+                sqlalchemy.sql.and_(table.columns.dataset_id.in_([ref.id for ref in refs]),
+                                    table.columns.datastore_name == datastoreName)
```

Comment:
At the moment this is only being called with a single ref and associated components so I don't think it's going to be unbounded. Of course in the future you might want to be doing bulk trashing rather than `for each ref: trash`.